app/models/assignment.rb

```diff
@@ -605,6 +605,65 @@ def summary_json(user)
       numMarked: self.get_num_marked(user.admin? ? nil : user.id) }
   end
 
+  # Generates the summary of the most test results associated with an assignment.
+  def summary_test_results
+    test_groups_query = self.test_groups
+                            .joins(test_group_results: { test_run: :grouping })
+                            .group('test_groups.id', 'groupings.id')
+                            .select('test_groups.id AS test_groups_id',
+                                    'MAX(test_group_results.created_at) AS test_group_results_created_at')
+                            .where.not('test_runs.submission_id': nil)
+                            .to_sql
+
+    TestGroup.joins(test_group_results: [:test_results, { test_run: { grouping: :group } }])
+             .joins("INNER JOIN (#{test_groups_query}) sub \
+                     ON test_groups.id = sub.test_groups_id \
+                     AND test_group_results_test_groups.created_at = sub.test_group_results_created_at")
+             .select('test_groups.name',
+                     'test_groups_id',
+                     'groups.group_name',
+                     'test_results.name as test_result_name',
+                     'test_results.status',
+                     'test_results.marks_earned',
+                     'test_results.marks_total',
+                     :output, :extra_info, :error_type)
+  end
+
+  # Generate a JSON summary of the most recent test results associated with an assignment.
+  def summary_test_result_json
+    self.summary_test_results.group_by(&:group_name).transform_values do |grouping|
+      grouping.group_by(&:name)
+    end.to_json
+  end
+
+  # Generate a CSV summary of the most recent test results associated with an assignment.
+  def summary_test_result_csv
+    summary_test_results = self.summary_test_results
+    results = summary_test_results.map do |result|
+      ["#{result['group_name']}:#{result['test_result_name']}", result['status']]
+    end.to_h
+
+    group_names = summary_test_results.group_by(&:group_name).keys
+
+    CSV.generate do |csv|
+      headers = SortedSet.new(results.keys)
+      csv << [nil, *headers]
+      group_names.each do |name|
+        row = [name]
+
+        headers.each do |header|
+          _, test_result_name = header.split(':')
```

Comment:
Don't split on colons, since there might actually be colons in the test name (this is configurable for some testers). But you won't need to do this if you change what I said above, since the `headers` here should exactly match the keys in the "inner" hashes in `results`.